### ISSUE TITLE
"Preparing Build Folder" is slow on my large git repo

### ISSUE THREAD

**Issue by jerrymarino** [buildkite/agent#734]
I'm using Buikdkite with a rather large git repo with a significant number years of history, contributors, and branches ( Pinterest iOS app ).

The majority of time spent in my jobs is preparing `Preparing Build Folder`. It spends nearly 10 minutes running `git clone`.

Before Buildkite I was doing `git pull` and incrementally loading the repo. Is there anything you recommend to git around the `git clone`?

**Comment by lox** [comment]
Hi! 👋🏻Big git repositories are certainly a pain! The agent will do a pull if there is already a repository checked out, you should only see a clone if it's the first time that machine has checked out your repository.

We've been discussing ways that we could speed this up, happy to work through some with you and your team! Can you drop us an email to hello@buildkite.com with an example slow build and we'll discuss it?

**Comment by buchgr** [comment]
@lox @jerrymarino We (Bazel) have started using `git clone --reference=<cache repository>` to avoid this problem. On each CI machine we have a docker container running that periodically fetches changes from the upstream repository to i.e. `/buildkite/cache-repo` and then on each CI build we do a `git clone --reference=/buildkite/cache-repo`. This reduces our clone size for a CI build from ~800MiB to a few KiB.

Hope that helps.

cc: @philwo

**Comment by toolmantim** [comment]
It might be neat to allow the `pre-checkout` hook to export a `BUILDKITE_AGENT_GIT_REFERENCE` env var which added that arg for you.

**Comment by jgavris** [comment]
We also see a fair number of clones during 'preparing build folder', despite our agents never explicitly clearing the repository. Having easier support for a reference clone would be awesome (as would figuring out who's blowing away the checkout!).

**Comment by jgavris** [comment]
At the same time, I wonder if the 'delete checkout dir on exit' could be more graceful if a build is interrupted due to a (force) push. If it were possible to let the `git checkout` phase finish gracefully on a cancel, that might avoid the 90% case and alleviate the need for implementing clones.

**Comment by jgavris** [comment]
We see this many times a day (a lot of re-checkouts of the entire repo) as our developers force push often to fix lint errors. There's no good reason I can think of that our repo needs to be checked out fully so often. We've made some mistakes in the past (committing large files before LFS was standard practice) so this hurts a lot.

**Comment by lox** [comment]
@jgavris would you be able to send us a build where it's incorrectly deleting the checkout?

**Comment by toolmantim** [comment]
Just for reference on this issue here, because it was talking about above and I didn't see this mentioned… you can use Git's [`--reference-if-able`](https://git-scm.com/docs/git-clone#git-clone---reference-if-ableltrepositorygt) flag by configuring the agent config property `git-clone-flags` (`BUILDKITE_GIT_CLONE_FLAGS`). You can also modify that config from an agent or plugin `pre-checkout` hook, by exporting the config env variable.

For example:

```
git-clone-flags="-v --reference-if-able /some/local/repo/checkout-1 --reference-if-able /some/local/repo/checkout-2"
```

**Comment by MisterRager** [comment]
Is there a way to use a shallow clone instead? I'm seeing checkout times approaching 7 minutes - by far the most expensive single part of any task.

**Comment by lox** [comment]
Git mirrors (reference bare repos) are available as an experiment in v3.10.0!

**Comment by jgavris** [comment]
Fantastic! Will try it out soon.
